### PR TITLE
Initial form state

### PR DESCRIPTION
**Pull request by gbirke**
It's now possible to initialize the form values with initial values
(which will be generated by the backend in a future commit).

Improved validation code to for easier handling of initial state.
Validation needs to know about initial state to allow for invalid values
as initial state (e.g. empty fields).

This is one step for task https://phabricator.wikimedia.org/T129227

### CODE REVIEW

File: app/js/lib/form_init.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+				self.components.forEach( function ( component ) {
+					component.render( formContent );
+				} );
+
+			} );
+		},
+		init: function ( initialFormValues ) {
+			this.connectValidatorsToStore( initialFormValues );
+			this.connectViewHandlersToStore();
+			this.connectComponentsToStore();
```
Comment (gbirke) [inline app/js/lib/form_init.js]: TODO: Trigger newInitializeValuesAction

File: app/js/lib/form_init.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+var reduxValidation = require( './redux_validation' ),
+	_ = require( 'lodash' ),
+	objectAssign = require( 'object-assign' ),
+
+	/**
+	 * Connect store updates to validators, components and view handlers
+	 */
+	FormEventConnector = {
```
Comment (gbirke) [inline app/js/lib/form_init.js]: Rename to StoreUpdateConnector

File: app/js/lib/form_init.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+var reduxValidation = require( './redux_validation' ),
+	_ = require( 'lodash' ),
+	objectAssign = require( 'object-assign' ),
+
+	/**
+	 * Connect store updates to validators, components and view handlers
```
Comment (gbirke) [inline app/js/lib/form_init.js]: Connect validators, components and view handlers to store updates

File: app/js/lib/form_init.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		},
+		init: function ( initialFormValues ) {
+			this.connectValidatorsToStore( initialFormValues );
+			this.connectViewHandlersToStore();
+			this.connectComponentsToStore();
+		}
+	};
+
+module.exports = {
+	initializeForm: function ( components, validatorFactoryFunction, viewHandlers, initialFormValues, store ) {
```
Comment (gbirke) [inline app/js/lib/form_init.js]: rename to initializeStoreUpdateHandling

File: app/js/main.js [hunk 5]
```diff
@@ -13,5 +13,6 @@ module.exports = {
 		createErrorBoxHandler:  require( './lib/view_handler/error_box' ).createHandler,
 		createFormPageVisibilityHandler: require( './lib/view_handler/form_page_visibility' ).createHandler
 	},
-	Actions: require( './lib/actions' )
+	Actions: require( './lib/actions' ),
+	initializeForm: require( './lib/form_init' ).initializeForm
```
Comment (gbirke) [inline app/js/main.js]: rename to connectSomething ....

File: app/js/tests/test_redux_validation.js [hunk 6]
```diff
@@ -74,6 +74,18 @@ test( 'createValidationDispatcher accepts validator object as validation functio
 	t.end();
 } );
 
+test( 'createValidationDispatcher can set initial values from object', function ( t ) {
```
Comment (gbirke) [inline app/js/tests/test_redux_validation.js]: Renam this test!!!  I'm confooz!

File: app/js/tests/test_redux_validation.js [hunk 7]
```diff
@@ -74,6 +74,18 @@ test( 'createValidationDispatcher accepts validator object as validation functio
...
+	var initialDataBag = { importantField: 'initial data', ignoredData: 'this won\'t be added to initial data' },
+		validationFunction = sinon.stub(),
+		actionCreationFunction = sinon.stub(),
+		dispatcher = reduxValidation.createValidationDispatcher( validationFunction, actionCreationFunction, [ 'importantField' ], initialDataBag ),
```
Comment (gbirke) [inline app/js/tests/test_redux_validation.js]: Line too long!

File: app/js/tests/test_redux_validation.js [hunk 8]
```diff
@@ -74,6 +74,18 @@ test( 'createValidationDispatcher accepts validator object as validation functio
...
+		expectedPreviousFieldValues = { importantField: 'initial data' };
+
+	t.deepEqual( dispatcher.previousFieldValues, expectedPreviousFieldValues, 'Previous field values should only contain selected values' );
```
Comment (gbirke) [inline app/js/tests/test_redux_validation.js]: too long!

File: app/js/tests/reducers/test_form_content.js [hunk 9]
```diff
@@ -42,5 +39,37 @@ test( 'CHANGE_CONTENT changes the field', function ( t ) {
...
+	} );
+	t.end();
 } );
+
+test( 'INITIALIZE_CONTENT changes multiple fields', function ( t ) {
+	var stateBefore = { paymentType: 'PPL', amount: 0, recurringPayment: 0 },
+		expectedState = { paymentType: 'BEZ', amount: '25,00', recurringPayment: 0 },
+		action = { type: 'INITIALIZE_CONTENT', payload: { amount: '25,00', paymentType: 'BEZ' } };
+
+	deepFreeze( stateBefore );
```
Comment (JeroenDeDauw) [inline app/js/tests/reducers/test_form_content.js]: Why is this needed?

File: app/js/tests/test_store_update_handling.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	return {
+		subscribe: sinon.spy(),
+		getState: sinon.stub().returns( storeData ),
+		// Call all callback functions that were registered through calls to `subscribe`
+		fakeUpdate: function () {
+			var subscribeCalls = this.subscribe.args;
+			subscribeCalls.forEach( function ( subscribeArguments ) {
+				// get the first argument to the `subscribe` call
+				var updateCallback = subscribeArguments[ 0 ];
+				// call the callback, using its own context
```
Comment (JeroenDeDauw) [inline app/js/tests/test_store_update_handling.js]: Does this comment add any value? Assuming you know what `call` does in JS, this ought to be obvious no?

File: app/js/tests/reducers/test_form_content.js (hunk 9 above)
Comment (gbirke) [inline app/js/tests/reducers/test_form_content.js]: To assert that the state is immutable. If the state is somehow changed in the reducer (instead of creating a new state object), the deep frozen object will throw an exception.
